plugins/outputs/loki/loki.go

```diff
@@ -0,0 +1,214 @@
+  # tls_key = "/etc/telegraf/key.pem"
+`
+
+type Loki struct {
+	Domain       string            `toml:"domain"`
+	Endpoint     string            `toml:"endpoint"`
+	Timeout      internal.Duration `toml:"timeout"`
+	Username     string            `toml:"username"`
+	Password     string            `toml:"password"`
+	Headers      map[string]string `toml:"headers"`
+	ClientID     string            `toml:"client_id"`
+	ClientSecret string            `toml:"client_secret"`
+	TokenURL     string            `toml:"token_url"`
+	Scopes       []string          `toml:"scopes"`
+	GZipRequest  bool              `toml:"gzip_request"`
+
+	url    string
+	client *http.Client
+	tls.ClientConfig
+}
+
+func (l *Loki) SampleConfig() string {
+	return sampleConfig
+}
+
+func (l *Loki) Description() string {
+	return "Send logs to Loki"
+}
+
+func (l *Loki) createClient(ctx context.Context) (*http.Client, error) {
+	tlsCfg, err := l.ClientConfig.TLSConfig()
+	if err != nil {
+		return nil, fmt.Errorf("tls config fail: %w", err)
+	}
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsCfg,
+			Proxy:           http.ProxyFromEnvironment,
+		},
+		Timeout: l.Timeout.Duration,
+	}
+
+	if l.ClientID != "" && l.ClientSecret != "" && l.TokenURL != "" {
+		oauthConfig := clientcredentials.Config{
+			ClientID:     l.ClientID,
+			ClientSecret: l.ClientSecret,
+			TokenURL:     l.TokenURL,
+			Scopes:       l.Scopes,
+		}
+		ctx = context.WithValue(ctx, oauth2.HTTPClient, client)
+		client = oauthConfig.Client(ctx)
+	}
+
+	return client, nil
+}
+
+func (l *Loki) Connect() (err error) {
+	if l.Domain == "" {
+		return fmt.Errorf("domain is required")
+	}
+
+	if l.Endpoint == "" {
+		l.Endpoint = defaultEndpoint
+	}
+
+	l.url = fmt.Sprintf("%s%s", l.Domain, l.Endpoint)
+
+	if l.Timeout.Duration == 0 {
+		l.Timeout.Duration = defaultClientTimeout
+	}
+
+	ctx := context.Background()
+	l.client, err = l.createClient(ctx)
+	if err != nil {
+		return fmt.Errorf("http client fail: %w", err)
+	}
+
+	return
+}
+
+func (l *Loki) Close() error {
+	l.client.CloseIdleConnections()
+
+	return nil
+}
+
+func (l *Loki) Write(metrics []telegraf.Metric) error {
+	s := Streams{}
+
+	for _, m := range metrics {
+		tags := m.TagList()
+		var line string
+
+		for _, f := range m.FieldList() {
+			line += fmt.Sprintf("%s=\"%v\" ", f.Key, f.Value)
+		}
+
+		s.insertLog(tags, Log{fmt.Sprintf("%d", m.Time().UnixNano()), line})
+	}
+
+	return l.write(s)
+}
+
+func (l *Loki) write(s Streams) error {
+	bs, err := json.Marshal(s)
+	if err != nil {
+		return fmt.Errorf("json.Marshal: %w", err)
+	}
+
+	var reqBodyBuffer io.Reader = bytes.NewBuffer(bs)
+
+	if l.GZipRequest {
+		rc, err := internal.CompressWithGzip(reqBodyBuffer)
+		if err != nil {
+			return err
+		}
+		defer rc.Close()
+		reqBodyBuffer = rc
+	}
+
+	req, err := http.NewRequest(http.MethodPost, l.url, reqBodyBuffer)
+	if err != nil {
+		return err
+	}
+
+	if l.Username != "" && l.Password != "" {
```

Comment:
This is also not quite correct, sorry! :-/ It is possible (at least for most services not sure about loki) to have a user name with empty password. So either we do
```suggestion
	if l.Username != "" {
```
or you check in `Connect()` that a username is given if the password is not empty and throw an error!